### PR TITLE
8316958: Add test for unstructured locking

### PR DESCRIPTION
**Pull request by rkennke**
In my original lightweight-locking PR, I had a testcase that exercises unstructured locking, which is permitted by the JVM spec, but not generated by javac, and thus not executed anywhere. JITs don't compile unstructured-locked code, therefore any code which attempts this would be slow anyway.

Testing:
 - [x] runtime/locking

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8316958](https://bugs.openjdk.org/browse/JDK-8316958): Add test for unstructured locking (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [d232eb26](https://git.openjdk.org/jdk/pull/15925/files/d232eb26a3464678b9d07c8380bf76f0cf4553e8)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15925/head:pull/15925` \
`$ git checkout pull/15925`

Update a local copy of the PR: \
`$ git checkout pull/15925` \
`$ git pull https://git.openjdk.org/jdk.git pull/15925/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15925`

View PR using the GUI difftool: \
`$ git pr show -t 15925`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15925.diff">https://git.openjdk.org/jdk/pull/15925.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15925#issuecomment-1735644980)

### CODE REVIEW

File: test/hotspot/jtreg/runtime/locking/TestUnstructuredLocking.jasm
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/locking/TestUnstructuredLocking.jasm:29]: I would add an Xcomp case to ensure the JIT really doesn't try to compile this, and that it still works.

File: test/hotspot/jtreg/runtime/locking/TestUnstructuredLocking.jasm
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/locking/TestUnstructuredLocking.jasm:50]: Curious why this is needed.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Good ides to have a test. I think we actually have a JNI-based unstructured locking testcase but not a bytecode one. One suggestion/request below.

Thanks

File: test/hotspot/jtreg/runtime/locking/TestUnstructuredLocking.jasm
Comment (rkennke) [reply test/hotspot/jtreg/runtime/locking/TestUnstructuredLocking.jasm:50]: I don't think it is needed. It's probably a left-over from an experiment that I did, maybe to avoid this whole method to be optimized-away (but that shouldn't matter, because it's not JIT-compiled anyway) or more likely to see how this interacts with inflation. In any case, I'll remove it.

File: test/hotspot/jtreg/runtime/locking/TestUnstructuredLocking.jasm
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/locking/TestUnstructuredLocking.jasm:32]: typo: camp instead of comp

**Review by dholmes-ora — APPROVED** [review]
One nit but otherwise looks good.
Thanks

**Review by dholmes-ora — APPROVED** [review]

**Comment by rkennke** [comment]
Should I wait for another review?

File: test/hotspot/jtreg/runtime/locking/TestUnstructuredLocking.jasm
Comment (shipilev) [inline test/hotspot/jtreg/runtime/locking/TestUnstructuredLocking.jasm:26]: Might want to call it `int`; this is what we do elsewhere.

File: test/hotspot/jtreg/runtime/locking/TestUnstructuredLocking.jasm
Comment (shipilev) [inline test/hotspot/jtreg/runtime/locking/TestUnstructuredLocking.jasm:32]: Similarly, might want to call it `comp`.

**Review by shipilev — APPROVED** [review]
Looks good to me, with minor nits.

(I am actually surprised one can make the pure jasm jtreg test, without any Java scaffolding, impressive.)

**Review by shipilev — APPROVED** [review]

**Pull request closed without merging**
